src/java.base/share/native/libjli/java.c

```diff
@@ -339,6 +339,46 @@ JLI_Launch(int argc, char ** argv,              /* main argc, argv */
 
     return JVMInit(&ifn, threadStackSize, argc, argv, mode, what, ret);
 }
+
+/*
+ * It's required to set or update LD_LIBRARY_PATH by the launcher in some cases:
+ * LD_LIBRARY_PATH contains libjvm.so of different version, linker doesn't
+ * support $ORIGIN or not able to solve dependency between libjvm and libjava.
+ * Runtime linker respects LD_LIBRARY_PATH variable only at the time of runtime
+ * image loading, so launcher has to re-execute itself if LD_LIBRARY_PATH
+ * was altered.
+ *
+ * All required logic is held within CreateExecutionEnvironment, this function is
+ * just a shell, that allows us to check for re-execution before any argument
+ * processing, environment variable or optionfile expansion.
+ */
+
+JNIEXPORT void JNICALL
+JLI_ReExecLauncher(int argc, char **argv) {
+
+    /**
+     * Fast path for the most common case
+     */
+#if !defined (MUSL_LIBC) && !defined(AIX)
+    if (getenv("LD_LIBRARY_PATH") == NULL) {
+        return;
+    }
+#endif
+
+    JLI_SetTraceLauncher();
```

Comment:
It's confusing to call JLI_SetTraceLauncher here as that is normally done by InitiLauncher.

Reply:
Agree. Will file a separate CR for java_md.c cleanup.